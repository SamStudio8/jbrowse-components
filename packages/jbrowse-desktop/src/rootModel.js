import { getSnapshot, types } from 'mobx-state-tree'
import { UndoManager } from 'mst-middlewares'
import JBrowseDesktop from './jbrowseModel'
import sessionModelFactory from './sessionModelFactory'

const { electronBetterIpc = {} } = window
const { ipcRenderer } = electronBetterIpc

export default function RootModel(pluginManager) {
  const Session = sessionModelFactory(pluginManager)
  return types
    .model('Root', {
      jbrowse: JBrowseDesktop(pluginManager, Session),
      session: types.maybe(Session),
      savedSessionNames: types.maybe(types.array(types.string)),
    })
    .actions(self => ({
      setSavedSessionNames(sessionNames) {
        self.savedSessionNames = sessionNames
      },
      setSession(sessionSnapshot) {
        self.session = sessionSnapshot
      },
      setDefaultSession() {
        self.setSession({
          ...self.jbrowse.defaultSession,
          name: `${self.jbrowse.defaultSession.name} ${new Date(
            Date.now(),
          ).toISOString()}`,
        })
      },
      renameCurrentSession(sessionName) {
        const snapshot = JSON.parse(JSON.stringify(getSnapshot(self.session)))
        const oldName = snapshot.name
        snapshot.name = sessionName
        self.setSession(snapshot)
        ipcRenderer.invoke('renameSession', oldName, sessionName)
      },
      duplicateCurrentSession() {
        const snapshot = JSON.parse(JSON.stringify(getSnapshot(self.session)))
        let newSnapshotName = `${self.session.name} (copy)`
        if (self.jbrowse.savedSessionNames.includes(newSnapshotName)) {
          let newSnapshotCopyNumber = 2
          do {
            newSnapshotName = `${self.session.name} (copy ${newSnapshotCopyNumber})`
            newSnapshotCopyNumber += 1
          } while (self.jbrowse.savedSessionNames.includes(newSnapshotName))
        }
        snapshot.name = newSnapshotName
        self.setSession(snapshot)
      },
      activateSession(sessionSnapshot) {
        self.setSession(sessionSnapshot)
        if (sessionSnapshot)
          self.setHistory(UndoManager.create({}, { targetStore: self.session }))
        else self.setHistory(undefined)
      },
      setHistory(history) {
        self.history = history
      },
    }))
    .volatile((/* self */) => ({
      history: {},
    }))
}
