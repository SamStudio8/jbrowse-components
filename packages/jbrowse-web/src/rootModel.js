import { getSnapshot, types } from 'mobx-state-tree'
import { UndoManager } from 'mst-middlewares'
import JBrowseWeb, { Session } from './jbrowseModel'

const RootModel = types
  .model('Root', {
    jbrowse: JBrowseWeb,
    session: types.maybe(Session),
  })
  .actions(self => ({
    setSession(sessionSnapshot) {
      self.session = sessionSnapshot
      self.jbrowse.updateSavedSession(sessionSnapshot)
    },
    setDefaultSession() {
      self.setSession({
        ...self.jbrowse.defaultSession,
        name: `${self.jbrowse.defaultSession.name} ${new Date(
          Date.now(),
        ).toISOString()}`,
      })
    },
    renameCurrentSession(sessionName) {
      const snapshot = JSON.parse(JSON.stringify(getSnapshot(self.session)))
      const oldName = snapshot.name
      snapshot.name = sessionName
      self.setSession(snapshot)
      self.jbrowse.replaceSavedSession(oldName, snapshot)
    },
    duplicateCurrentSession() {
      const snapshot = JSON.parse(JSON.stringify(getSnapshot(self.session)))
      let newSnapshotName = `${self.session.name} (copy)`
      if (self.jbrowse.savedSessionNames.includes(newSnapshotName)) {
        let newSnapshotCopyNumber = 2
        do {
          newSnapshotName = `${self.session.name} (copy ${newSnapshotCopyNumber})`
          newSnapshotCopyNumber += 1
        } while (self.jbrowse.savedSessionNames.includes(newSnapshotName))
      }
      snapshot.name = newSnapshotName
      self.setSession(snapshot)
    },
    activateSession(name) {
      const newSessionSnapshot = self.jbrowse.savedSessions.find(
        sessionSnap => sessionSnap.name === name,
      )
      if (!newSessionSnapshot)
        throw new Error(
          `Can't activate session ${name}, it is not in the savedSessions`,
        )
      self.setSession(newSessionSnapshot)
    },
    setHistory(history) {
      self.history = history
    },
  }))
  .volatile((/* self */) => ({
    history: {},
  }))

export function createTestSession(snapshot = {}) {
  const root = RootModel.create({
    jbrowse: {
      configuration: { rpc: { defaultDriver: 'MainThreadRpcDriver' } },
    },
  })
  root.setSession({
    name: 'testSession',
    menuBars: [{ type: 'MainMenuBar' }],
    ...snapshot,
  })
  root.setHistory(UndoManager.create({}, { targetStore: root.session }))
  return root.session
}

export default RootModel
