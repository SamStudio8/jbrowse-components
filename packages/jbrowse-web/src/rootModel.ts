import PluginManager from '@gmod/jbrowse-core/PluginManager'
import { getSnapshot, types, cast, SnapshotIn } from 'mobx-state-tree'
import { UndoManager } from 'mst-middlewares'
import corePlugins from './corePlugins'
import jbrowseWebFactory from './jbrowseModel'
import sessionModelFactory from './sessionModelFactory'

// eslint-disable-next-line @typescript-eslint/no-explicit-any
export default function RootModel(pluginManager: any) {
  const Session = sessionModelFactory(pluginManager)
  return types
    .model('Root', {
      jbrowse: jbrowseWebFactory(pluginManager, Session),
      session: types.maybe(Session),
    })
    .actions(self => ({
      setSession(sessionSnapshot: SnapshotIn<typeof Session>) {
        self.session = cast(sessionSnapshot)
        self.jbrowse.updateSavedSession(sessionSnapshot)
      },
      setDefaultSession() {
        this.setSession({
          ...self.jbrowse.defaultSession,
          name: `${self.jbrowse.defaultSession.name} ${new Date(
            Date.now(),
          ).toISOString()}`,
        })
      },
      renameCurrentSession(sessionName: string) {
        if (self.session) {
          const snapshot = JSON.parse(JSON.stringify(getSnapshot(self.session)))
          const oldName = snapshot.name
          snapshot.name = sessionName
          self.jbrowse.replaceSavedSession(oldName, snapshot)
          this.setSession(snapshot)
        }
      },
      duplicateCurrentSession() {
        if (self.session) {
          const snapshot = JSON.parse(JSON.stringify(getSnapshot(self.session)))
          let newSnapshotName = `${self.session.name} (copy)`
          if (self.jbrowse.savedSessionNames.includes(newSnapshotName)) {
            let newSnapshotCopyNumber = 2
            do {
              newSnapshotName = `${self.session.name} (copy ${newSnapshotCopyNumber})`
              newSnapshotCopyNumber += 1
            } while (self.jbrowse.savedSessionNames.includes(newSnapshotName))
          }
          snapshot.name = newSnapshotName
          this.setSession(snapshot)
        }
      },
      activateSession(name: string) {
        const newSessionSnapshot = self.jbrowse.savedSessions.find(
          sessionSnap => sessionSnap.name === name,
        )
        if (!newSessionSnapshot)
          throw new Error(
            `Can't activate session ${name}, it is not in the savedSessions`,
          )
        this.setSession(newSessionSnapshot)
      },
    }))
    .volatile((/* self */) => ({
      history: {},
    }))
    .actions(self => ({
      // eslint-disable-next-line @typescript-eslint/no-explicit-any
      setHistory(history: any) {
        self.history = history
      },
    }))
}

export function createTestSession(snapshot = {}) {
  const pluginManager = new PluginManager(corePlugins.map(P => new P()))
  pluginManager.createPluggableElements()

  const JBrowseRootModel = RootModel(pluginManager)
  const root = JBrowseRootModel.create({
    jbrowse: {
      configuration: { rpc: { defaultDriver: 'MainThreadRpcDriver' } },
    },
  })
  root.setSession({
    name: 'testSession',
    ...snapshot,
  })
  root.setHistory(UndoManager.create({}, { targetStore: root.session }))
  pluginManager.setRootModel(root)

  pluginManager.configure()
  return root.session
}
