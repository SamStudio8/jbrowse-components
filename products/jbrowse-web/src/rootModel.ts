import assemblyManagerFactory, {
  assemblyConfigSchemas as AssemblyConfigSchemasFactory,
} from '@jbrowse/core/assemblyManager'
import PluginManager from '@jbrowse/core/PluginManager'
import RpcManager from '@jbrowse/core/rpc/RpcManager'
import { MenuItem } from '@jbrowse/core/ui'
import { AbstractSessionModel } from '@jbrowse/core/util'
import AddIcon from '@material-ui/icons/Add'
import {
  addDisposer,
  cast,
  getSnapshot,
  SnapshotIn,
  types,
} from 'mobx-state-tree'
import { observable, autorun } from 'mobx'
import { UndoManager } from 'mst-middlewares'
import * as uuid from 'uuid'
import corePlugins from './corePlugins'
import jbrowseWebFactory from './jbrowseModel'
// @ts-ignore
import RenderWorker from './rpc.worker'
import sessionModelFactory from './sessionModelFactory'

interface Menu {
  label: string
  menuItems: MenuItem[]
}

export default function RootModel(
  pluginManager: PluginManager,
  adminMode = false,
) {
  const Session = sessionModelFactory(pluginManager)
  const { assemblyConfigSchemas, dispatcher } = AssemblyConfigSchemasFactory(
    pluginManager,
  )
  const assemblyConfigSchemasType = types.union(
    { dispatcher },
    ...assemblyConfigSchemas,
  )
  const assemblyManagerType = assemblyManagerFactory(assemblyConfigSchemasType)
  return types
    .model('Root', {
      jbrowse: jbrowseWebFactory(
        pluginManager,
        Session,
        assemblyConfigSchemasType,
      ),
      session: types.maybe(Session),
      assemblyManager: assemblyManagerType,
      error: types.maybe(types.string),
      version: types.maybe(types.string),
    })
    .volatile(() => ({
      savedSessionsVolatile: observable.map({}),
    }))
    .views(self => ({
      get savedSessions() {
        return Array.from(self.savedSessionsVolatile.values())
      },
    }))
    .views(self => ({
      get savedSessionNames() {
        return self.savedSessions.map(session => session.name)
      },
      get currentSessionId() {
        const locationUrl = new URL(window.location.href)
        const params = new URLSearchParams(locationUrl.search)
        return params?.get('session')?.split('local-')[1]
      },
      get hasRecoverableAutosave() {
        return !!Object.keys(localStorage).find(
          key => key === 'localSaved-previousAutosave',
        )
      },
      get isUnsavedSession() {
        const locationUrl = new URL(window.location.href)
        const params = new URLSearchParams(locationUrl.search)
        return params?.get('session')?.startsWith('local-')
      },
    }))
    .actions(self => ({
      afterCreate() {
        Object.entries(localStorage)
          .filter(
            ([key, _val]) =>
              key.startsWith('localSaved-') &&
              !key.endsWith('previousAutosave'),
          )
          .forEach(([key, val]) => {
            self.savedSessionsVolatile.set(key, JSON.parse(val).session)
          })

        addDisposer(
          self,
          autorun(() => {
            for (const [key, val] of self.savedSessionsVolatile.entries()) {
              try {
                localStorage.setItem(`${key}`, JSON.stringify({ session: val }))
              } catch (e) {
                if (e.code === '22' || e.code === '1024') {
                  // eslint-disable-next-line no-alert
                  alert(
                    'Local storage is full! Please use the "Open sessions" panel to remove old sessions',
                  )
                }
              }
            }
          }),
        )
      },
      setSession(sessionSnapshot?: SnapshotIn<typeof Session>) {
        self.session = cast(sessionSnapshot)
      },
      setDefaultSession() {
        const newSession = {
          ...self.jbrowse.defaultSession,
          name: `${self.jbrowse.defaultSession.name} ${new Date(
            Date.now(),
          ).toLocaleDateString()} ${new Date(Date.now()).toLocaleTimeString()}`,
        }

        const localId = `local-${uuid.v4()}`
        sessionStorage.clear()
        sessionStorage.setItem(localId, JSON.stringify({ session: newSession }))
        this.setSessionUuidInUrl(localId)
        this.setSession(newSession)
        return localId
      },
      renameCurrentSession(sessionName: string) {
        if (self.session) {
          const snapshot = JSON.parse(JSON.stringify(getSnapshot(self.session)))
          snapshot.name = sessionName
          this.setSession(snapshot)
        }
      },
      duplicateCurrentSession() {
        if (self.session) {
          const snapshot = JSON.parse(JSON.stringify(getSnapshot(self.session)))
          let newSnapshotName = `${self.session.name} (copy)`
          if (self.savedSessionNames.includes(newSnapshotName)) {
            let newSnapshotCopyNumber = 2
            do {
              newSnapshotName = `${self.session.name} (copy ${newSnapshotCopyNumber})`
              newSnapshotCopyNumber += 1
            } while (self.savedSessionNames.includes(newSnapshotName))
          }
          snapshot.name = newSnapshotName
          const localId = `local-${uuid.v4()}`
          sessionStorage.clear()
          sessionStorage.setItem(localId, JSON.stringify({ session: snapshot }))
          this.setSessionUuidInUrl(localId)
          this.setSession(snapshot)
        }
      },
      activateSession(name: string) {
        const newSessionSnapshot = localStorage.getItem(`localSaved-${name}`)
        if (!newSessionSnapshot)
          throw new Error(
            `Can't activate session ${name}, it is not in the savedSessions`,
          )

        const [localId, sessionObj] = newSessionSnapshot
        this.setSessionUuidInUrl(localId)
        this.setSession(JSON.parse(sessionObj).session)
      },
      saveSessionToLocalStorage() {
        if (self.session && self.isUnsavedSession) {
          self.savedSessionsVolatile.set(
            `localSaved-${self.session.name}`,
            getSnapshot(self.session),
          )
          // const snapshot = JSON.parse(JSON.stringify(getSnapshot(self.session)))
          // // @ts-ignore
          // const localId = `localSaved-${self.session.name}`
          //
        }
      },
      loadAutosaveSession() {
        const autosavedSession = JSON.parse(
          localStorage.getItem('localSaved-previousAutosave') || '',
        ).session
        autosavedSession.name = `${autosavedSession.name.replace(
          '-autosaved',
          '',
        )}-restored`
        const localId = `local-${uuid.v4()}`
        sessionStorage.clear()
        sessionStorage.setItem(
          localId,
          JSON.stringify({ session: autosavedSession }),
        )
        this.setSessionUuidInUrl(localId)
        this.setSession(autosavedSession)
        return localId
      },
      setSessionUuidInUrl(localId: string) {
        const locationUrl = new URL(window.location.href)
        const params = new URLSearchParams(locationUrl.search)
        params.set('session', `${localId}`)
        locationUrl.search = params.toString()
        window.history.replaceState({}, '', locationUrl.href)
      },
      setError(errorMessage: string) {
        self.error = errorMessage
      },
    }))
    .volatile(self => ({
      history: {},
      menus: [
        {
          label: 'File',
          menuItems: [
            {
              label: 'New session',
              icon: AddIcon,
              // eslint-disable-next-line @typescript-eslint/no-explicit-any
              onClick: (session: any) => {
                const lastAutosave = localStorage.getItem('autosave')
                if (lastAutosave) {
                  localStorage.setItem(
                    'localSaved-previousAutosave',
                    lastAutosave,
                  )
                }
                session.setDefaultSession()
              },
            },
          ],
        },
      ] as Menu[],
      rpcManager: new RpcManager(
        pluginManager,
        self.jbrowse.plugins,
        self.jbrowse.configuration.rpc,
        {
          WebWorkerRpcDriver: { WorkerClass: RenderWorker },
          MainThreadRpcDriver: {},
        },
      ),
      adminMode,
    }))
    .actions(self => ({
      // eslint-disable-next-line @typescript-eslint/no-explicit-any
      setHistory(history: any) {
        self.history = history
      },
      setMenus(newMenus: Menu[]) {
        self.menus = newMenus
      },
      /**
       * Add a top-level menu
       * @param menuName - Name of the menu to insert.
       * @returns The new length of the top-level menus array
       */
      appendMenu(menuName: string) {
        return self.menus.push({ label: menuName, menuItems: [] })
      },
      /**
       * Insert a top-level menu
       * @param menuName - Name of the menu to insert.
       * @param position - Position to insert menu. If negative, counts from th
       * end, e.g. `insertMenu('My Menu', -1)` will insert the menu as the
       * second-to-last one.
       * @returns The new length of the top-level menus array
       */
      insertMenu(menuName: string, position: number) {
        const insertPosition =
          position < 0 ? self.menus.length + position : position
        self.menus.splice(insertPosition, 0, { label: menuName, menuItems: [] })
        return self.menus.length
      },
      /**
       * Add a menu item to a top-level menu
       * @param menuName - Name of the top-level menu to append to.
       * @param menuItem - Menu item to append.
       * @returns The new length of the menu
       */
      appendToMenu(menuName: string, menuItem: MenuItem) {
        const menu = self.menus.find(m => m.label === menuName)
        if (!menu) {
          self.menus.push({ label: menuName, menuItems: [menuItem] })
          return 1
        }
        return menu.menuItems.push(menuItem)
      },
      /**
       * Insert a menu item into a top-level menu
       * @param menuName - Name of the top-level menu to insert into
       * @param menuItem - Menu item to insert
       * @param position - Position to insert menu item. If negative, counts
       * from the end, e.g. `insertMenu('My Menu', -1)` will insert the menu as
       * the second-to-last one.
       * @returns The new length of the menu
       */
      insertInMenu(menuName: string, menuItem: MenuItem, position: number) {
        const menu = self.menus.find(m => m.label === menuName)
        if (!menu) {
          self.menus.push({ label: menuName, menuItems: [menuItem] })
          return 1
        }
        const insertPosition =
          position < 0 ? menu.menuItems.length + position : position
        menu.menuItems.splice(insertPosition, 0, menuItem)
        return menu.menuItems.length
      },
      /**
       * Add a menu item to a sub-menu
       * @param menuPath - Path to the sub-menu to add to, starting with the
       * top-level menu (e.g. `['File', 'Insert']`).
       * @param menuItem - Menu item to append.
       * @returns The new length of the sub-menu
       */
      appendToSubMenu(menuPath: string[], menuItem: MenuItem) {
        let topMenu = self.menus.find(m => m.label === menuPath[0])
        if (!topMenu) {
          const idx = this.appendMenu(menuPath[0])
          topMenu = self.menus[idx - 1]
        }
        let { menuItems: subMenu } = topMenu
        const pathSoFar = [menuPath[0]]
        menuPath.slice(1).forEach(menuName => {
          pathSoFar.push(menuName)
          let sm = subMenu.find(mi => 'label' in mi && mi.label === menuName)
          if (!sm) {
            const idx = subMenu.push({ label: menuName, subMenu: [] })
            sm = subMenu[idx - 1]
          }
          if (!('subMenu' in sm)) {
            throw new Error(
              `"${menuName}" in path "${pathSoFar}" is not a subMenu`,
            )
          }
          subMenu = sm.subMenu
        })
        return subMenu.push(menuItem)
      },
      /**
       * Insert a menu item into a sub-menu
       * @param menuPath - Path to the sub-menu to add to, starting with the
       * top-level menu (e.g. `['File', 'Insert']`).
       * @param menuItem - Menu item to insert.
       * @param position - Position to insert menu item. If negative, counts
       * from the end, e.g. `insertMenu('My Menu', -1)` will insert the menu as
       * the second-to-last one.
       * @returns The new length of the sub-menu
       */
      insertInSubMenu(
        menuPath: string[],
        menuItem: MenuItem,
        position: number,
      ) {
        let topMenu = self.menus.find(m => m.label === menuPath[0])
        if (!topMenu) {
          const idx = this.appendMenu(menuPath[0])
          topMenu = self.menus[idx - 1]
        }
        let { menuItems: subMenu } = topMenu
        const pathSoFar = [menuPath[0]]
        menuPath.slice(1).forEach(menuName => {
          pathSoFar.push(menuName)
          let sm = subMenu.find(mi => 'label' in mi && mi.label === menuName)
          if (!sm) {
            const idx = subMenu.push({ label: menuName, subMenu: [] })
            sm = subMenu[idx - 1]
          }
          if (!('subMenu' in sm)) {
            throw new Error(
              `"${menuName}" in path "${pathSoFar}" is not a subMenu`,
            )
          }
          subMenu = sm.subMenu
        })
        subMenu.splice(position, 0, menuItem)
        return subMenu.length
      },
    }))
}

export function createTestSession(snapshot = {}) {
  const pluginManager = new PluginManager(corePlugins.map(P => new P()))
  pluginManager.createPluggableElements()

  const JBrowseRootModel = RootModel(pluginManager)
  const root = JBrowseRootModel.create({
    jbrowse: {
      configuration: { rpc: { defaultDriver: 'MainThreadRpcDriver' } },
    },
    assemblyManager: {},
  })
  root.setSession({
    name: 'testSession',
    ...snapshot,
  })
  root.setHistory(UndoManager.create({}, { targetStore: root.session }))
  // @ts-ignore
  root.session.views.map(view => view.setWidth(800))
  pluginManager.setRootModel(root)

  pluginManager.configure()
  return root.session as AbstractSessionModel
}
