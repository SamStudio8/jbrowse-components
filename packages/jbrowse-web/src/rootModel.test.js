import { getSnapshot } from 'mobx-state-tree'
import rootModel from './rootModel'

describe('Root MST model', () => {
  it('creates with defaults', () => {
    const root = rootModel.create({
      jbrowse: {
        configuration: { rpc: { defaultDriver: 'MainThreadRpcDriver' } },
      },
    })
    expect(root.session).toBeUndefined()
    expect(root.jbrowse.savedSessions.length).toBe(0)
    root.setDefaultSession()
    expect(root.session).toBeTruthy()
    expect(root.jbrowse.savedSessions.length).toBe(1)
    expect(root.jbrowse.datasets.length).toBe(0)
    expect(getSnapshot(root.jbrowse.configuration)).toMatchSnapshot({
      configId: expect.any(String),
      rpc: { configId: expect.any(String) },
    })
  })

  it('creates with a minimal session', () => {
    const root = rootModel.create({
      jbrowse: {
        configuration: { rpc: { defaultDriver: 'MainThreadRpcDriver' } },
      },
      session: { name: 'testSession' },
    })
    expect(root.session).toBeTruthy()
  })

  it('activates a session snapshot', () => {
    const root = rootModel.create({
      jbrowse: {
        configuration: { rpc: { defaultDriver: 'MainThreadRpcDriver' } },
        savedSessions: [{ name: 'testSession' }],
      },
    })
    expect(root.session).toBeUndefined()
    root.setSession(root.jbrowse.savedSessions[0])
    expect(root.session).toBeTruthy()
  })

  it('adds track and connection configs to a dataset', () => {
    const root = rootModel.create({
      jbrowse: {
        configuration: { rpc: { defaultDriver: 'MainThreadRpcDriver' } },
        datasets: [
          {
            name: 'dataset1',
            assembly: { name: 'assembly1', aliases: ['assemblyA'] },
          },
        ],
      },
    })
    expect(root.jbrowse.datasets.length).toBe(1)
    expect(getSnapshot(root.jbrowse.datasets[0])).toMatchSnapshot({
      configId: expect.any(String),
      assembly: {
        configId: expect.any(String),
      },
    })
    const newTrackConf = root.jbrowse.datasets[0].addTrackConf({
      type: 'BasicTrack',
    })
    expect(getSnapshot(newTrackConf)).toMatchSnapshot({
      configId: expect.any(String),
      adapter: {
        configId: expect.any(String),
        index: { configId: expect.any(String) },
      },
      renderer: { configId: expect.any(String) },
    })
    expect(root.jbrowse.datasets[0].tracks.length).toBe(1)
    const newConnectionConf = root.jbrowse.datasets[0].addConnectionConf({
      type: 'JBrowse1Connection',
    })
    expect(getSnapshot(newConnectionConf)).toMatchSnapshot({
      configId: expect.any(String),
    })
    expect(root.jbrowse.datasets[0].connections.length).toBe(1)
  })

  it('adds a session snapshot', () => {
    const root = rootModel.create({
      jbrowse: {
        configuration: { rpc: { defaultDriver: 'MainThreadRpcDriver' } },
      },
    })
    root.jbrowse.addSavedSession({ name: 'testSession' })
    expect(root.jbrowse.savedSessions.length).toBe(1)
  })

  it('throws if session is invalid', () => {
    expect(() =>
      rootModel.create({
        jbrowse: {
          configuration: { rpc: { defaultDriver: 'MainThreadRpcDriver' } },
        },
        session: {},
      }),
    ).toThrow()
  })

  it('throws if session snapshot is invalid', () => {
    expect(() =>
      rootModel.create({
        jbrowse: {
          configuration: { rpc: { defaultDriver: 'MainThreadRpcDriver' } },
          savedSessions: [{}],
        },
      }),
    ).toThrow()
  })
})
