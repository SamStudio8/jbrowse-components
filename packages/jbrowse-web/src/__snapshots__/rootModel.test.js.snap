// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Root MST model adds track and connection configs to a dataset 1`] = `
Object {
  "assembly": Object {
    "aliases": Array [
      "assemblyA",
    ],
    "configId": Any<String>,
    "name": "assembly1",
  },
  "configId": Any<String>,
  "name": "dataset1",
}
`;

exports[`Root MST model adds track and connection configs to a dataset 2`] = `
Object {
  "adapter": Object {
    "configId": Any<String>,
    "index": Object {
      "configId": Any<String>,
    },
    "type": "BamAdapter",
  },
  "configId": Any<String>,
  "renderer": Object {
    "configId": Any<String>,
    "type": "PileupRenderer",
  },
  "type": "BasicTrack",
}
`;

exports[`Root MST model adds track and connection configs to a dataset 3`] = `
Object {
  "configId": Any<String>,
  "type": "JBrowse1Connection",
}
`;

exports[`Root MST model creates with defaults 1`] = `
Object {
  "configId": Any<String>,
  "rpc": Object {
    "configId": Any<String>,
    "defaultDriver": "MainThreadRpcDriver",
  },
}
`;
